alf/algorithms/data_transformer.py

```diff
@@ -819,32 +816,96 @@ def transform_experience(self, experience: Experience):
+                                 future_idx).unsqueeze(1))
 
             # relabel desired goal
             result_desired_goal = alf.nest.get_field(result,
                                                      self._desired_goal_field)
-            relabed_goal = result_desired_goal.clone()
+            relabeled_goal = result_desired_goal.clone()
             her_batch_index_tuple = (her_indices.unsqueeze(1),
                                      torch.arange(batch_length).unsqueeze(0))
-            relabed_goal[her_batch_index_tuple] = future_ag
+            if has_her:
+                relabeled_goal[her_batch_index_tuple] = future_ag[her_indices]
 
             # recompute rewards
             result_ag = alf.nest.get_field(result, self._achieved_goal_field)
-            relabeled_rewards = self._reward_fn(result_ag, relabed_goal)
+            relabeled_rewards = self._reward_fn(
+                result_ag, relabeled_goal, threshold=self._threshold)
+            if alf.summary.should_record_summaries():
+                alf.summary.scalar(
+                    "replayer/" + buffer._name +
+                    ".discount_mean_before_relabel",
+                    torch.mean(result.discount[:, 1:]))
+            if self._sparse_reward:
+                reward_achieved = relabeled_rewards >= 0
+                # Cut off episode for any goal reached.
+                end = reward_achieved
+                discount = torch.where(end, torch.tensor(0.), result.discount)
+                step_type = torch.where(end, torch.tensor(StepType.LAST),
+                                        result.step_type)
+                # Also relabel ``LAST``` steps to ``MID``` where aux goals were not
+                # achieved but env ended episode due to position goal achieved.
+                # -1/0 reward doesn't end episode on achieving position goal, and
+                # doesn't need to do this relabeling.
+                goal_reward = result.reward
+                if len(result.reward.shape) > 2:
+                    goal_reward = result.reward[..., 0]
+                mid = (
+                    result.step_type == StepType.LAST) & ~reward_achieved & (
+                        goal_reward > 0)  # assumes no multi dim goal reward.
+                discount = torch.where(mid, torch.tensor(1.), discount)
+                step_type = torch.where(mid, torch.tensor(StepType.MID),
+                                        step_type)
+
+                if alf.summary.should_record_summaries():
+                    alf.summary.scalar(
+                        "replayer/" + buffer._name +
+                        ".discount_mean_after_relabel",
+                        torch.mean(discount[:, 1:]))
+
+                result = result._replace(discount=discount)
+                result = result._replace(step_type=step_type)
+                relabeled_rewards = suite_socialbot.transform_reward_tensor(
```

Comment:
My understanding is that `HindsightExperienceTransformer` is a general data transformer independent of the environment. So it feels strange to include `suite_socialbot` here.